### ISSUE TITLE
Error: Invalid for_each argument for aws data subnets

### ISSUE THREAD

**Issue by nick4fake** [hashicorp/terraform#26878]
### Terraform Version

```
0.13.5
```

### Terraform Configuration Files

```terraform
data "aws_vpc" "this" {
  filter {
    name   = "tag:Name"
    values = [var.name]
  }
}

data "aws_subnet_ids" "private" {
  vpc_id = data.aws_vpc.this.id

  dynamic "filter" {
    for_each = var.private_subnets_tags
    content {
      name   = "tag:${filter.key}"
      values = [filter.value]
    }
  }
}

data "aws_subnet_ids" "public" {
  vpc_id = data.aws_vpc.this.id

  dynamic "filter" {
    for_each = var.public_subnets_tags
    content {
      name   = "tag:${filter.key}"
      values = [filter.value]
    }
  }
}

data "aws_subnet" "private" {
  for_each = data.aws_subnet_ids.private.ids

  id = each.value
}

data "aws_subnet" "public" {
  for_each = data.aws_subnet_ids.public.ids

  id = each.value
}
```

### Crash Output

````
 Error: Invalid for_each argument
# 
#   on main.tf line 33, in data "aws_subnet" "private":
#   33:   for_each = data.aws_subnet_ids.private.ids
# 
# The "for_each" value depends on resource attributes that cannot be determined
# until apply, so Terraform cannot predict how many instances will be created.
# To work around this, use the -target argument to first apply only the
# resources that the for_each depends on.
````
### Expected Behavior

Data sources work.

### Actual Behavior

Somehow data-only code is still not enough for plan

### Additional Context

Please keep in mind that ALL code is static - it only has variables and data sources, so it is not quite clear what attributes are unknown until apply.

### References

https://github.com/hashicorp/terraform/issues/24348

**Comment by jbardin** [comment]
Hi @nick4fake 

Without a complete configuration its hard to determine what is actually going on here. If either `var.private_subnets_tags` or `var.name` is unknown, then the data sources cannot be read, and cannot be used as `for_each` values. 

Can you show how you are using this section of the configuration?

**Comment by nick4fake** [comment]
@jbardin 

I think you might have misunderstood me. I use this module **directly**. There are no other resources at all, all inputs are static.

Anyway, variables.tf
````terraform
variable "managed" {
  type = bool
}

variable "name" {
  type = string
}

variable "private_subnets_tags" {
  type = map(string)
}

variable "public_subnets_tags" {
  type = map(string)
}
````

Inputs are provided as file with.

**Comment by nick4fake** [comment]
@jbardin that's why I am confused. How can I see that error if I only have data sources? Everything should be available before apply as there is nothing to even apply, I only use outputs:

````terraform
output "name" {
  value = data.aws_vpc.this.tags["Name"]
}

output "vpc_id" {
  value = data.aws_vpc.this.id
}

output "arn" {
  value = data.aws_vpc.this.arn
}

output "cidr" {
  value = data.aws_vpc.this.cidr_block
}

output "private_subnets_ids" {
  value = [for v in data.aws_subnet.private : v.id]
}

output "private_subnets_arns" {
  value = [for v in data.aws_subnet.private : v.arn]
}

output "private_subnets_cidrs" {
  value = [for v in data.aws_subnet.private : v.cidr_block]
}

output "public_subnets_ids" {
  value = [for v in data.aws_subnet.public : v.id]
}

output "public_subnets_arns" {
  value = [for v in data.aws_subnet.public : v.arn]
}

output "public_subnets_cidrs" {
  value = [for v in data.aws_subnet.public : v.cidr_block]
}
````

**Comment by jbardin** [comment]
Unfortunately I'm not able to replicate the error with the information provided, there must be something else in the configuration or environment that we're not seeing here. The value `data.aws_subnet_ids.private.ids` is unknown at the time of planning, which indicates that at least one of the data sources could not be read before the plan was attempted. 

If you could provide the trace logs and/or a reproducible example we may be able to determine what's missing here.
If you want to test this with the latest 0.14 release, it may give us more information about what is going on (or even fix the issue), since terraform now handles data source refreshing entirely within the planning process.

Thanks!

**Comment by nick4fake** [comment]
@jbardin I'll give v0.14.0-rc1 a try.

**Comment by innovate-invent** [comment]
I am having the same issue with a different provider:
```hcl
resource "galaxy_repository" "islandcompare" {
  for_each = { for repo in data.galaxy_workflow_repositories.islandcompare.repositories: "${repo.tool_shed}/repos/${repo.owner}/${repo.name}" => repo }
  tool_shed = each.value.tool_shed
  owner = each.value.owner
  name = each.value.name
  changeset_revision = each.value.changeset_revision
}
```

```
Error: Invalid for_each argument

  on ../../destinations/galaxy/tools.tf line 6, in resource "galaxy_repository" "islandcompare":
   6:   for_each = { for repo in data.galaxy_workflow_repositories.islandcompare.repositories: "${repo.tool_shed}/repos/${repo.owner}/${repo.name}" => repo }

```

despite having data.galaxy_workflow_repositories.islandcompare.repositories already loaded into the state file:
```
> terraform state show module.islandcompare.module.galaxy.data.galaxy_workflow_repositories.islandcompare
# module.islandcompare.module.galaxy.data.galaxy_workflow_repositories.islandcompare:
data "galaxy_workflow_repositories" "islandcompare" {
    id           = "1V5hJz74ndrVNhBUnzfFd5Zw9WI="
    json         = "1V5hJz74ndrVNhBUnzfFd5Zw9WI="
    repositories = [
        {
            changeset_revision = "22bf5b566ce2"
            name               = "parsnp"
            owner              = "brinkmanlab"
            tool_shed          = "toolshed.g2.bx.psu.edu"
        },
        {
            changeset_revision = "2889433c7ae1"
            name               = "ncbi_blast_plus"
            owner              = "devteam"
            tool_shed          = "toolshed.g2.bx.psu.edu"
        },
...
```

**Comment by nick4fake** [comment]
@jbardin upgrading to 0.14-rc1 helped (so far), thanks.

Do you know when it is going to be released?

**Comment by innovate-invent** [comment]
I compiled from the v0.14 branch to resolve the unrelated error and I now still get this one.

**Comment by jbardin** [comment]
@nick4fake we do not have an official release date, it largely depends on output of testing with rc1. 

@innovate-invent, Unfortunately we cannot tell why the data source is not able to be read during plan from the information provided. The logs may provide some help, but `data.galaxy_workflow_repositories.islandcompare.repositories` is unknown during plan, which is resulting in the error seen.

**Comment by innovate-invent** [comment]
[log.txt](https://github.com/hashicorp/terraform/files/5550227/log.txt)

What is really bizarre is I use this module in another deployment and it works fine. The only thing I can imagine is different is the module/resource path.

**Comment by nick4fake** [comment]
@jbardin not sure if it helps without my log, but original issue with data sources was only present during destroy when resources have already been deleted.

**Comment by jbardin** [comment]
Thanks @nick4fake, that more or less confirms it as a refresh problem, which is fixed in 0.14.

@innovate-invent, Sorry, but I can't tell what's going on here without the configuration. We can't tell _why_ the `galaxy_workflow_repositories` data source is unknown without knowing all the inputs. While this has a similar error, it appears to be a different issue. If you can supply all the necessary information to reproduce the error, go ahead and file a new issue.

Thanks!
